Add loop bound state registry

kelunik opened the pull request:
The registry allows to store information bound to a specific loop. This can be useful to store state of globals like a DNS resolver.

**Rationale for a global DNS resolver**

Nobody wants to inject a resolver into everything that does some sort of I/O. Using IP addresses everywhere instead of hostnames also doesn't make sense.

Therefore there should be a global default DNS resolver. But this resolver needs a read watcher for the response. If the loop gets swapped and a DNS request is made, the resolver would never get the read event, since the stream is only watched in the old loop. The resolver also doesn't get any info about the loop swap, so it can't use a new watcher.

With a loop bound registry, it can just store the connection in the loop and will use a new connection when the loop is swapped, because the state doesn't exist any longer. Once the loop is put back, it can use the old connection again.

This API should only be used by real global objects. Most packages won't need to use this API, but it's required for the mentioned use cases.

kelunik committed:
Add loop bound state registry

The registry allows to store information bound to a specific loop.
This can be useful to store state of globals like a DNS resolver.

Rationale for a global DNS resolver:

Nobody wants to inject a resolver into everything that does some sort of I/O.
Using IP addresses everywhere instead of hostnames also doesn't make sense.

Therefore there should be a global default DNS resolver. But this resolver
needs a read watcher for the response. If the loop gets swapped and a DNS
request is made, the resolver would never get the read event, since the
stream is only watched in the old loop. The resolver also doesn't get any
info about the loop swap, so it can't use a new watcher.

With a loop bound registry, it can just store the connection in the loop and
will use a new connection when the loop is swapped, because the state doesn't
exist any longer. Once the loop is put back, it can use the old connection
again.

This API should only be used by real global objects. Most packages won't need
to use this API, but it's required for the mentioned use cases.

bwoebi commented:
Looks good to me; I'd like to have someone else merge it though.

kelunik commented:
I don't think we have consensus here yet, or do we?

kelunik commented:
Not sure whether `storeState` / `putState` / `setState`. I don't like `setState`, it makes me think it's an event loop state, not something not actually related to the loop driver but just data.

bwoebi commented:
@kelunik yeah, I'm not sure on consensus, that's why I'd like to have someone else merging it when it's clear that there's consensus … as I'm biased as it was my idea ;-)

bwoebi commented:
@kelunik either we have get/set or store/fetch. Both variations are fine with me (but not set/fetch).

assertchris commented:
What about putting the registry stuff into a trait, or a subclass? Subclass may be slower than desired. Is anything else likely to need this registry stuff?

bwoebi commented:
Having a trait to do that... hmm... (You mean additionally to the interface methods, right?)

assertchris commented:
@bwoebi yeah, in addition.

kelunik commented:
What's the benefit of moving it out when just the `Loop` uses it?

assertchris commented:
It's additional responsibility, that could be tested and documented on its own.

kelunik commented:
The registry has to be swapped with the loop, how does that work with a trait?

assertchris commented:
Traits are literally compiler-assisted copy and paste. It would work in exactly the same way as if the code was inside the loop class.

kelunik commented:
In that case the trait would rely on a variable being present in the loop class. I don't see the benefit here. Just a disadvantage that we rely on a static variable being present in the trait.

Separation of concerns is fine, but not for the sake of it if you have to use a trait for it. Because that doesn't separate anything, it just moves it out, so it can be used in multiple places.

assertchris commented:
If the property is on the trait then the class would rely on the variable being present. That's an ok relationship since the class already knows it's using the trait. Alternatively, there are abstract method signatures to ensure traits have access to the data from the classes that use them.

They're separate responsibilities. I won't bore you with the benefits of SRP (there are good books which do a much better job of explaining it than I do), but essentially one trades speed for isolated testing, documentation, maintenance and limited side effects to modifying behaviour.

kelunik commented:
Yes, but this is a utility class that just has static methods and only `execute` changes state at all, so the number of site effects is quite limited.

> If the property is on the trait then the class would rely on the variable being present.

Traits can't define static members in PHP, otherwise I'd probably be fine with it.

bwoebi commented:
@kelunik Why should traits not be able to define static members? Works in 5.5 all the way up to current master.

@assertchris Thinking about it, I don't think traits are a very good way to enforce SRP. Please leave it in. I'd like to have one class with all the public methods viewable at once. A trait only hides these.
In general, it _is_ fine with SRP. The responsibility is managing the loop, and part of it is managing the attached state too.

The PR is LGTM to me and shall be merged.

kelunik commented:
@bwoebi Seems like the German translation is just plain wrong then.

> Innerhalb von Traits kann auf statische Variablen zugegriffen werden, aber diese können nicht in Traits definiert werden. Statische Methoden können hingegen auch durch Traits für die darstellende Klasse definiert werden. http://php.net/manual/de/language.oop5.traits.php

bwoebi commented:
@kelunik well… Fixed in doc commit 339170.

kelunik committed:
Refactor Registry into a trait, so it can be tested separately

kelunik commented:
@assertchris I think it's fine as trait then. Refactored it and added tests.

It ensures we can test the registry without an event loop / outside of `Loop::execute`.

kelunik committed:
Fixup registry reset after execution of the loop

kelunik committed:
Change namespace of Registry

kelunik committed:
Add test to autoload-dev

AndrewCarterUK commented on src/Registry.php:
```diff
@@ -0,0 +1,57 @@
+
+namespace Interop\Async;
+
+trait Registry {
```
Opening brace should be on a new line according to lord FIG.

kelunik replied:
Done.

AndrewCarterUK commented on test/RegistryTest.php:
```diff
@@ -0,0 +1,60 @@
+
+namespace Interop\Async;
+
+class RegistryTest extends \PHPUnit_Framework_TestCase {
```
Opening brace should be on a new line according to lord FIG (as with all the methods here).

kelunik replied:
Done.

kelunik committed:
Add phpunit to require-dev again

kelunik committed:
Apply ugly FIG braces style

kelunik commented:
Can we merge this?

kelunik commented:
@bwoebi @trowski @AndrewCarterUK Can you merge this? I don't want to merge my own PR.
